### ISSUE TITLE
.NETCORE 2.0 Lambda Invocation throws "Could not find the required deps.json."

### ISSUE THREAD

**Issue by mipnw** [aws/aws-lambda-dotnet#243]
I deployed a [.NET CORE 2.0](https://aws.amazon.com/blogs/developer/aws-lambda-net-core-2-0-support-released/) C# project to AWS Lambda using the [AWS Toolkit for Visual Studio 2017](https://marketplace.visualstudio.com/items?itemName=AmazonWebServices.AWSToolkitforVisualStudio2017).

The AWS Toolkit, upon successfully completing the deployment, opens a Visual Studio tab to test the invocation of the AWS Lambda. I wrote a sample JSON request and clicked on "Invoke".

The result was
{
    "errorType" : "LambdaException",
    "errorMessage" : **"Could not find the required 'deps.json'.  This file should be present at the root of the deployment package."**
}

I have no idea what this deps.json is, it looks like a bug in the publishing or perhaps in the runtime.

This looks similar to [Issue #119](https://github.com/aws/aws-lambda-dotnet/issues/119)

**Comment by mipnw** [comment]
I believe I found the bug.

The AWS Toolkit for Visual Studio 2017, when it publishes, produces a file named <function-name>.deps.json, where <function-name> is the name of the function the user specifies via attribute "function-name" in aws-lambda-tools-defaults.json.

But the AWS Lambda runtime looks for a file named <Namespace>.deps.json where <Namespace> is the C# namespace the user specified via attribute "function-handler" in aws-lambda-tools-defaults.json.

When the C# namespace is not identical to the lambda function name, the lambda fails to execute.

**Comment by mipnw** [comment]
Nevermind, I misunderstood how to properly format attribute `function-handler` in `aws-lambda-tools-defaults.json`. It's `<assembly>::<namespace.type>::<function>`, not `<namespace>::<type>::<function>`.

So one can indeed have an assembly name that is not identical to one's namespace, without hitting this FileNotFound exception.

**Comment by fschroder** [comment]
I got a similar error message, but in my case the problem was in the Handler property of serverless.template. Pointing to the wrong assembly produced an error about a missing deps.json file.

**Comment by fillic2002** [comment]
so do we need to modify function-handler with <assembly>::<namespace.type>::<function>? if so i am seeing in my case its formatted as "AWSLambda3::AWSLambda3.Function::FunctionHandler" which looks right to me but still i am seeing error.

**Comment by vellozzi** [comment]
Check to see if the deps.json file is in the zip file that's getting deployed.  Also double check that ```<GenerateRuntimeConfigurationFiles>true</GenerateRuntimeConfigurationFiles>``` is set in your csproj file.

**Comment by fillic2002** [comment]
I tried adding the same file where the issue is appearing. But no change in error....

**Comment by 1000106** [comment]
hi guys i am also facing same issue as per "fillic2002"
I am keeping following files in zip folder( Zip folder name -> AWSLambda2.zip)
1. AWSLambda2.deps.json
2. AWSLambda2.dll
3. AWSLambda2.runtimeconfig.dev.json
4. AWSLambda2.runtimeconfig.json

After uploading this zip folder when i am trying to test my AWSLambda2 function getting following error

  "errorType": "LambdaException",
  "errorMessage": "Could not find the required 'AWSLambda2.deps.json'.  This file should be present at the root of the deployment package."

Function handler as per below mention
 "function-handler": "AWSLambda2::AWSLambda2.Function::FunctionHandler"

can any one have solution on it..??

**Comment by 1000106** [comment]
@fillic2002  , do you get any solution...??

**Comment by taryel-hlontsi** [comment]
Hello
have the same problem.
My zip package contains only two JSON files:

- <Namespace>.deps.json
- <Namespace>.runtimeconfig.json

When deploying via VS2017 everything is ok, however when create a lambda function in AWS console and upload zip package manually it throws the following error if I try to test it:

`Could not find the required 'LambdaTest.deps.json'. This file should be present at the root of the deployment package.: LambdaException`

Why does it want LambdaTest.deps.json?
Could someone advise please?

**Comment by 1000106** [comment]
hey @taryel-hlontsi 
1. Don't do package manually ( if you are doing this then you have to add "LambdaTest.deps.json" file explicitly from your project root folder) and while creating zip file do't zip file folder . Go inside folder select all files and then zip with your project name.
2. You can also zip your package from visual studio command prompt.
3.  Publish your project from Visual studio itself by right clicking solution. ( This is the best option)

**Comment by patoncrispy** [comment]
I'm having issues with this, too. I have bundled and deployed my `bin/Debug/netcoreapp2.1/publish/` which contains my `*.deps.json`. I have downloaded and verified that the deployed zip file contains the `*.deps.json`, too, but I'm still getting this error. I've checked that it's not a typo. My `AWS::Serverless::Function` handler has the correct Assembly name, namespace and function using the convention for an `APIGatewayProxyFunction`:  `"Assembly::Name.Space.LambdaEntryPoint::FunctionHandlerAsync"`. I'm not sure what could be causing this issue as everything I've done seems correct. :(

**Comment by webedge** [comment]
Same issue. Downloaded and verified that the file exists and that the path is correct. I'm deploying a .NET Core API and all binaries along with the CloudFormation template are at the root of the ZIP package. Still no go. Any suggestions?

**Comment by vellozzi** [comment]
Does your deps.json file match your handler?
If the handler is:
```
Assembly::Namespace.Class::Method
```
the deps.json file needs to be named:
```
Assembly.deps.json
```

If it doesn't, then check out https://github.com/aws/aws-lambda-dotnet/issues/313 which includes a workaround.
If it does match let us know.

**Comment by patoncrispy** [comment]
Mine matches.

**Comment by vellozzi** [comment]
Are all of your files in the root of your zip file?

**Comment by patoncrispy** [comment]
Yes.

**Comment by patoncrispy** [comment]
Maybe this issue should be re-opened?

**Comment by dsab123** [comment]
I can publish my project to Lambda using Visual Studio, but the manual upload method does not seem to work. I'm guessing that there's an issue with how the project gets packaged. I'd like to pull down the code from the AWS console to compare, but there doesn't seem to be an option to do so..

**Comment by dsab123** [comment]
I stand corrected, I _can_ pull down the function.

**Comment by patoncrispy** [comment]
This is the output of the zip generated by the SAM CLI. I've modified my project assembly name but you'll see that both the dll and deps.json are present.

AWSSDK.Core.dll
AWSSDK.Extensions.NETCore.Setup.dll
Amazon.Lambda.APIGatewayEvents.dll
Amazon.Lambda.AspNetCoreServer.dll
Amazon.Lambda.Core.dll
Amazon.Lambda.Logging.AspNetCore.dll
Amazon.Lambda.Serialization.Json.dll
MyProject.Assembly.deps.json
MyProject.Assembly.dll
MyProject.Assembly.pdb
MyProject.Assembly.runtimeconfig.json
GraphQL-Parser.dll
GraphQL.dll
GraphiQL.dll
Microsoft.CodeAnalysis.CSharp.Workspaces.dll
Microsoft.CodeAnalysis.Workspaces.dll
Microsoft.VisualStudio.Web.CodeGeneration.Contracts.dll
Microsoft.VisualStudio.Web.CodeGeneration.Core.dll
Microsoft.VisualStudio.Web.CodeGeneration.EntityFrameworkCore.dll
Microsoft.VisualStudio.Web.CodeGeneration.Templating.dll
Microsoft.VisualStudio.Web.CodeGeneration.Utils.dll
Microsoft.VisualStudio.Web.CodeGeneration.dll
Microsoft.VisualStudio.Web.CodeGenerators.Mvc.dll
MySqlConnector.dll
NuGet.Frameworks.dll
Pomelo.EntityFrameworkCore.MySql.dll
Pomelo.JsonObject.dll
System.Composition.AttributedModel.dll
System.Composition.Convention.dll
System.Composition.Hosting.dll
System.Composition.Runtime.dll
System.Composition.TypedParts.dll
System.Reactive.Core.dll
System.Reactive.Interfaces.dll
System.Reactive.Linq.dll
appsettings.Development.json
appsettings.json
aws-lambda-tools-defaults.json
dotnet-aspnet-codegenerator-design.dll
event.json
web.config

**Comment by vellozzi** [comment]
Can I also see the handler string that was used to create the function?

**Comment by patoncrispy** [comment]
My `template.yaml` has:
```
Resources:
  API:
    Type: AWS::Serverless::Function
    Properties:
      Handler: MyProject.Assembly::MyProject.Assembly.LambdaEntryPoint::FunctionHandlerAsync
      Runtime: dotnetcore2.1
      CodeUri: dist/deploy.zip
```

And I have a `build.sh` file looks like:

```
#!/usr/bin/env bash

echo "Cleaning..."
dotnet clean -v q

echo "Publishing..."
dotnet publish -c Release -v q

mkdir dist
rm -rf dist/*

zip -j dist/deploy.zip bin/Release/netcoreapp2.1/publish/*

sam package \
--template-file template.yaml \
--s3-bucket my-deployments \
--output-template-file template-output.yaml

sam deploy \
--template-file template-output.yaml \
--stack-name mystackname  \
--capabilities CAPABILITY_IAM
```

**Comment by loop-evgeny** [comment]
I'm also running into this. I exported the lambda package from AWS and verified that the file it's complaining about is there.

**Comment by patoncrispy** [comment]
Has it broken since?

**Comment by loop-evgeny** [comment]
It only breaks - sometimes - when I deploy a new version of my lambda function. If I redeploy it a few times it works again and then it keeps working as long as I leave it alone.

**Comment by patoncrispy** [comment]
That's no good! I still haven't got mine working. :(

**Comment by patoncrispy** [comment]
@normj  is there any chance this issue could be investigated further?

**Comment by loop-evgeny** [comment]
Well, now the error happens every time, no how much I redeploy the function. My Lambda function is unusable. This is awful!

**Comment by fillic2002** [comment]
> Well, now the error happens every time, no how much I redeploy the function. My Lambda function is unusable. This is awful!

Are you deploying via VS? through publish lambda or via AWS portal?

**Comment by loop-evgeny** [comment]
Deploying by running `dotnet lambda deploy-function --profile MY_PROFILE_NAME --disable-interactive true` with Amazon.Lambda.Tools version 2.2.0

**Comment by fillic2002** [comment]
> Deploying by running `dotnet lambda deploy-function --profile MY_PROFILE_NAME --disable-interactive true` with Amazon.Lambda.Tools version 2.2.0

what role or policies applied on this lambda? This you can check by login into AWS portal

**Comment by patoncrispy** [comment]
> what role or policies applied on this lambda? This you can check by login into AWS portal

How could roles or policies affect a file not being detected in the deployment package?

**Comment by loop-evgeny** [comment]
I also don't see how this could affect it, but the Lambda function uses a role especially defined for it, which has access to some DynamoDB tables that it uses and another policy to allow access logs:

```
            "Effect": "Allow",
            "Action": [
                "logs:CreateLogGroup",
                "logs:CreateLogStream",
                "logs:PutLogEvents"
            ],
            "Resource": "arn:aws:logs:*:*:*"
```

This also hasn't changed in a while.

**Comment by fillic2002** [comment]
> > what role or policies applied on this lambda? This you can check by login into AWS portal
> 
> How could roles or policies affect a file not being detected in the deployment package?
 you have a point.... lets look at the json file you have setup in project

**Comment by fillic2002** [comment]
> > > what role or policies applied on this lambda? This you can check by login into AWS portal
> > 
> > 
> > How could roles or policies affect a file not being detected in the deployment package?
> > you have a point.... lets look at the json file you have setup in project

Hope your json file name is "aws-lambda-tools-defaults" and i am interested in the various settings in there...

**Comment by loop-evgeny** [comment]
Yes, it is named "aws-lambda-tools-defaults.json". It looks like this:

```

{
    "Information" : [
        "This file provides default values for the deployment wizard inside Visual Studio and the AWS Lambda commands added to the .NET Core CLI.",
        "To learn more about the Lambda commands with the .NET Core CLI execute the following command at the command line in the project root directory.",
        "dotnet lambda help",
        "All the command line options for the Lambda command can be specified in this file."
    ],
    "profile"     : "SOME_OTHER_PROFILE",
    "region"      : "eu-central-1",
    "configuration" : "Release",
    "framework"     : "netcoreapp2.0",
    "function-runtime" : "dotnetcore2.0",
    "function-memory-size" : 128,
    "function-timeout"     : 15,
    "function-handler"     : "MY_ASSEMBLY_NAME::MY_NAMESPACE.Function::FunctionHandler",
    "function-name"        : "MY_LAMBDA_NAME",
    "function-role"        : "arn:aws:iam::MY_ACCOUNT_ID:role/MY_LAMBDA_ROLE",
    "tracing-mode"         : "PassThrough",
    "environment-variables" : ""
}
```

**Comment by fillic2002** [comment]
> > "function-handler" : "MY_ASSEMBLY_NAME::MY_NAMESPACE.Function::FunctionHandler",
> 
> can you please check this detail for me if its setup correctly?

Also have you tried publishing this lambda from solution explorer itself?

**Comment by loop-evgeny** [comment]
It's correct and has not changed in a long time. I haven't used VS to publish the Lambda recently and don't even have the tools installed any more.

**Comment by patoncrispy** [comment]
I've been using the SAM CLI, so `aws-lambda-tools-defaults.json` shouldn't be relevant. Also, `function-handler` doesn't even seem to be a [required field](https://github.com/aws-samples/aws-net-guides/blob/master/Serverless/Serverless%20App%20with%20Dynamo-Example/TestServerless/aws-lambda-tools-defaults.json).

**Comment by loop-evgeny** [comment]
It seems that this problem happens whenever I run the deployment (`dotnet lambda deploy-function`) on a Linux machine and never when I run it on a Windows machine!

**Comment by vellozzi** [comment]
We're looking at this.  

It's difficult to sort out because there are different reasons why the deps.json file might not be found.  So each person's situation could be different.

**Comment by vellozzi** [comment]
We've added a fallback mechanism for the deps.json file search to Lambda in .NET Core 2.1.  

Once deployed, Lambda will first look for Assembly.deps.json (where Assembly is from the handler: Assembly::Namespace.Class::Method).  If Assembly.deps.json isn't found it will look to see if there is **one and only one** deps.json in the root of the deployment package.  If that's the case it will use that deps.json file.

I'll update this issue when the change is deployed.  In the meantime we'd recommend upgrading your functions to use .NET Core 2.1 so you can take advantage of the new fallback mechanism.

**Comment by loop-evgeny** [comment]
Thank you, but it would be good if you could get to the bottom of why this happens consistently when the deployment is run from Linux, even though the file still exists in the deployment package. The fallback mechanism is, at best, a workaround and doesn't address the real cause of this.

**Comment by patoncrispy** [comment]
I agree. A workaround will help us move on but if there are things that perhaps we aren't doing right then it would be good to know.

**Comment by vellozzi** [comment]
@patoncrispy - There at least one situation we know of where users aren't doing anything wrong, and the deps.json file still isn't found.  That's what the fix is for.  Ideally we'd like to find the root cause to every case.

@loop-evgeny - that's one case I'd like to look into separately.

**Comment by patoncrispy** [comment]
@vellozzi sounds good. :)

**Comment by vellozzi** [comment]
The fallback mechanism for the ```.deps.json``` file has been released to production.  Remember that it only works in .NET Core 2.1 Lambda functions.  

Please upgrade your functions to .NET Core 2.1 and try it out.  Let us know if you're still having trouble.

**Comment by drvora** [comment]
> The fallback mechanism for the `.deps.json` file has been released to production. Remember that it only works in .NET Core 2.1 Lambda functions.
> 
> Please upgrade your functions to .NET Core 2.1 and try it out. Let us know if you're still having trouble.

We are receiving this error even when deploying to Lambda .NET core 2.1.

**Comment by manu-one** [comment]
Getting the same error:
framework : netcoreapp2.1

Scenarios tested;

- Visual Studio 2017 - UI : Publishing works, Trying out the Lambda Function also works.
- Ubuntu 16 WSL - CLI  : Publishing works, Trying to execute throws error "No deps.json" found.
- PowerShell - CLI  : Publishing works, Trying to execute throws error "No deps.json" found.
- CMD - CLI  : Publishing works, Trying to execute throws error "No deps.json" found.

If I try again with the same code, Visual Studio UI method will still do a good job. No errors.

I'm working on a SAM template and multiple Lambda Functions.

This is a current and replicable issue, please let us know if there is a workaround that can allow us to use CLI gain.

Thanks.

**Comment by CheeryLittlebottom** [comment]
Had the same issue, noticed the file permissions within the zip were set to 000. Changing this solved the issue for me.

**Comment by TryTryAgain** [comment]
This should be reopened. I've come to the conclusion this may be related to building the package on Linux for me... Permissions changing/umask stuff didn't help. Would love to be able to use this but it just isn't working.

**Comment by dude0001** [comment]
Having the same issue as @CheeryLittlebottom
https://github.com/aws/aws-lambda-dotnet/issues/243#issuecomment-477153221

When building the assembly on a Linux build agent, it would not work. But worked fine from a Windows build agent.  Issue ended up being permissions of the files in the .ZIP. Need to chmod -r 755 before zipping if your build agent is Linux.

**Comment by vijayadandamudi22** [comment]
AWS is not able to restore the dependency packages.
First try to check the if the function-handler name matches.
Then try publishing with powershell or CLI using command "dotnet publish -o dist".
Once published use this folder to create the zip file. This resolved the issue.

**Comment by JordanKelley** [comment]
I was having this issue whenever loading a packaged up lambda function zip file, and the thing that solved this error for me was making sure the zipped folder name was the same as the assembly/namespace. For some reason I was initially uploading a zipped file with the name "publish" instead of my lambda function name.

**Comment by DanielCarns** [comment]
I was also having this same issue after renaming the default .NET SAM components. I tried many of these work arounds but none seemed to fix it. What ended up resolving my problem was using the --guided flag when deploying. The only difference in deployments that I observed was that it didn't confirm the changeset. Hope this helps someone!

**Comment by yuriusV** [comment]
In my case, LF line endings (instead of CR LF) are the reason of error

**Comment by MickBadran** [comment]
Hey folks - I know this is a little old, but I had this exact same issue with the latest AWS tools and VS2019 -> even a 'publish to AWS Lambda' didnt work.

Works fine in local debug mode.

When deployed to a lambda function - the function runtime complains about *.deps.json & *.runtimeconfig.json not found.

My project is created straight off the template solutions that come with AWS tools - nothing fancy.
I noticed when publishing, rather than just grab the 'output dir' the task wants to zip the entire project from the root of the project. (typically where the *.csproj file is).

So I added a couple of post-build event lines to copy those files to the project root.

Worked a treat.
